### PR TITLE
add `unicode` scalar function

### PR DESCRIPTION
**Pull request by ethanniser**
rewrite of #173 

the emoji integration test doesnt pass but the unit test does I cant figure out why

### CODE REVIEW

**Comment by penberg** [comment]
The emoji issue may be a parser issue where it turns UTF-8 into ASCII, for example.

File: core/types.rs [hunk 1]
```diff
@@ -42,7 +42,7 @@ impl Display for OwnedValue {
             OwnedValue::Integer(i) => write!(f, "{}", i),
             OwnedValue::Float(fl) => write!(f, "{:?}", fl),
             OwnedValue::Text(s) => write!(f, "{}", s),
-            OwnedValue::Blob(b) => write!(f, "{:?}", b),
+            OwnedValue::Blob(b) => write!(f, "{}", String::from_utf8_lossy(b)),
```
Comment (ethanniser) [inline core/types.rs:45]: also this is dependent on pragma encoding I think? probably fine for now

**Comment by ethanniser** [comment]
@penberg the issue is with the testing framework somehow because running the tests with `sqlite3` has the same issue and running `SELECT unicode('😊');` in the limbo shell works as expected as well

File: testing/scalar-functions.test [hunk 2]
```diff
@@ -243,3 +243,26 @@ do_execsql_test date-with-invalid-timezone {
...
+  SELECT unicode('a');
+} {97}
+
+do_execsql_test unicode-emoji {
+  SELECT unicode('😊');
+} {128522}
+
+do_execsql_test unicode-empty {
+  SELECT unicode('');
+} {NULL}
```
Comment (penberg) [inline testing/scalar-functions.test:256]: This test looks incorrect:

```
sqlite> SELECT unicode('');

```

**Comment by penberg** [comment]
@ethanniser I don't see any issue with the test framework either, I just think that the test you wrote for `unicode('')` is incorrect.

File: testing/scalar-functions.test (hunk 2 above)
Comment (ethanniser) [reply testing/scalar-functions.test:256]: image
in what way?

Comment (ethanniser) [reply testing/scalar-functions.test]: nvn

**Comment by ethanniser** [comment]
its definitely the tests

image

image


image

**Comment by penberg** [comment]
@ethanniser Can you please fold all the commits into one to clean up git history a bit and press the "ready for review" button so we can get this thing merged?

**Comment by ethanniser** [comment]
sure, but there is still this outstanding test issue

im happy to just delete the one that doesnt pass if you really want to merge

**Comment by ethanniser** [comment]
wait all the tests pass ci??

**Comment by ethanniser** [comment]
must be a local issue for me then very strange I think this should be good to merge if it looks good to you

thanks again for the help

**Pull request merged**